boa_gc/src/lib.rs

```diff
@@ -222,141 +248,153 @@ impl Allocator {
         unsafe {
-            Self::sweep(&gc.adult_start, &mut gc.runtime.bytes_allocated);
+            Self::sweep(
+                &gc.strong_start,
+                &gc.weak_start,
+                &mut gc.runtime.bytes_allocated,
+            );
         }
     }
 
     /// Walk the heap and mark any nodes deemed reachable
-    fn mark_heap(head: &Cell<Option<NonNull<GcBox<dyn Trace>>>>) -> Vec<NonNull<GcBox<dyn Trace>>> {
+    fn mark_heap(
+        mut strong: &Cell<Option<NonNull<GcBox<dyn Trace>>>>,
+        mut weak: &Cell<Option<NonNull<dyn ErasedEphemeronBox>>>,
+    ) -> Unreachables {
         let _timer = Profiler::global().start_event("Gc Marking", "gc");
         // Walk the list, tracing and marking the nodes
-        let mut finalize = Vec::new();
-        let mut ephemeron_queue = Vec::new();
-        let mut mark_head = head;
-        while let Some(node) = mark_head.get() {
-            // SAFETY: node must be valid as it is coming directly from the heap.
+        let mut strong_dead = Vec::new();
+        let mut pending_ephemerons = Vec::new();
+
+        // === Preliminary mark phase ===
+        //
+        // 0. Get the naive list of possibly dead nodes.
+        while let Some(node) = strong.get() {
+            // SAFETY: node must be valid as this phase cannot drop any node.
             let node_ref = unsafe { node.as_ref() };
-            if node_ref.header.is_ephemeron() {
-                ephemeron_queue.push(node);
-            } else if node_ref.header.roots() > 0 {
+            if node_ref.header.roots() > 0 {
                 // SAFETY: the reference to node must be valid as it is rooted. Passing
                 // invalid references can result in Undefined Behavior
                 unsafe {
-                    node_ref.trace_inner();
+                    node_ref.mark_and_trace();
                 }
-            } else {
-                finalize.push(node);
+            } else if !node_ref.is_marked() {
+                strong_dead.push(node);
             }
-            mark_head = &node_ref.header.next;
+            strong = &node_ref.header.next;
         }
 
-        // Ephemeron Evaluation
-        if !ephemeron_queue.is_empty() {
-            ephemeron_queue = Self::mark_ephemerons(ephemeron_queue);
+        // 0.1. Early return if there are no ephemerons in the GC
+        if weak.get().is_none() {
+            strong_dead.retain_mut(|node| {
+                // SAFETY: node must be valid as this phase cannot drop any node.
+                unsafe { !node.as_ref().is_marked() }
+            });
+            return Unreachables {
+                strong: strong_dead,
+                weak: Vec::new(),
+            };
         }
 
-        // Any left over nodes in the ephemeron queue at this point are
-        // unreachable and need to be notified/finalized.
-        finalize.extend(ephemeron_queue);
-
-        finalize
-    }
+        // === Weak mark phase ===
```

Comment:
Apologies about the delay on this. I still need to try and do some testing on this branch, but I thought I'd leave a review with some feedback/what I'm looking at. 

A lot of this looks good to me, but I'm a little concerned here on the weak mark phase since this is adding all ephemerons to the queue by default. It definitely would be a better approach than the thread_local queue if it works, but I'm worried this may end up holding onto some objects longer than expected or creating an immortal self-referencing chain. Would it be possible to add more test cases involving a chain of weak and non-weak items?

Reply:
Added some tests to check the correct behaviour on self-references